File: models/pywatershed/scripts/append_seg_data.py
import os
import numpy  as np
import pandas as pd
from netCDF4 import Dataset

# Define file paths
# Change the file path. Ideally, these files should live in the 'work' directory
filepath = '/Users/gharamti/Documents/MATLAB/Derecho/HydroDART/pywatershed/'
table_f, source_f, dest_f = 'river_midpoints.csv', 'parameters_dis_seg.nc', 'parameters_dis_seg_app.nc'
NTWRK_FILE, PARAM_FILE, CSV_TABLE = map(lambda f: os.path.join(filepath, f), [dest_f, source_f, table_f])

if not os.path.isfile(PARAM_FILE) or not os.path.isfile(CSV_TABLE):
    raise FileNotFoundError(f"Missing file: {PARAM_FILE} or {CSV_TABLE}")

# Read CSV and sort by index
# Could read headers to check for the right columns
data = pd.read_csv(CSV_TABLE).iloc[:, [4, 5, 6]].values
data = data[np.argsort(data[:, 0])] # sorted array

lats, lons = data[:, 1], data[:, 2]

# Read NetCDF dimensions & variables
with Dataset(PARAM_FILE, 'r') as nc:
    n_seg, gages         = len(nc.dimensions['nsegment']), len(nc.dimensions['npoigages'])
    to_index, poi_gauges = nc.variables['tosegment'][:].astype(int), nc.variables['poi_gage_id'][:].astype(int)

# Compute upstream links
fromIndices, num_up_links, fromIndsStart, fromIndsEnd = [], np.zeros(n_seg, int), np.zeros(n_seg, int), np.zeros(n_seg, int)

k = 0
for i in range(n_seg):
    upstream = np.where(to_index == (i + 1))[0] + 1  # Convert to 1-based index
    num_up_links[i] = len(upstream)
    if upstream.size > 0:
        fromIndsStart[i], fromIndsEnd[i] = k + 1, k + len(upstream)
        fromIndices.extend(upstream)
        k += len(upstream)

# Create new NetCDF file
if os.path.isfile(NTWRK_FILE):
    os.remove(NTWRK_FILE)

with Dataset(NTWRK_FILE, 'w', format='NETCDF4') as nc_new, Dataset(PARAM_FILE, 'r') as nc_old:
    # Copy dimensions
    for dim_name, dim in nc_old.dimensions.items():
        nc_new.createDimension(dim_name, len(dim) if not dim.isunlimited() else None)
    nc_new.createDimension('index', len(fromIndices))  

    # Copy variables
    for var_name, var in nc_old.variables.items():
        new_var    = nc_new.createVariable(var_name, var.datatype, var.dimensions)
        new_var[:] = var[:]

    # Ignore attributes for now

    # Create new variables
    nc_new.createVariable('fromIndsStart', 'i8', ('nsegment',))[:]  = fromIndsStart
    nc_new.createVariable('fromIndsEnd'  , 'i8', ('nsegment',))[:]  = fromIndsEnd
    nc_new.createVariable('fromIndices'  , 'i8', ('index',))[:]     = np.array(fromIndices, int)
    nc_new.createVariable('num_up_links' , 'i8', ('nsegment',))[:]  = num_up_links
    nc_new.createVariable('seg_lats'     , 'f8', ('nsegment',))[:]  = lats
    nc_new.createVariable('seg_lons'     , 'f8', ('nsegment',))[:]  = lons
    nc_new.createVariable('poi_gauges'   , 'i8', ('npoigages',))[:] = poi_gauges

File: models/pywatershed/scripts/append_seg_data.m
function append_seg_data

% Change the path 'filepath' to netcdf files. 
% PARAM_FILE : Original 'parameters_dis_seg.nc' file
% CSV_TABLE  : Includes segment data from the shapefile 'river_midpoints.csv'
% NTWRK_FILE : Similar to 'parameters_dis_seg.nc' but with appended variables 

% The following variables are created and appended
% to the parameters_dis_seg.nc geometry file:
%
% - seg_lats      : Latitude of the individual segments (at midpoint?)
% - seg_lons      : Longitude of the individual segments 
% - poi_gauges    : USGS ID for the available gauges  
% - num_up_links  : Number of upstream segments at each reach
% - fromIndsStart : Index of first upstream segment
% - fromIndsEnd   : Index of last upstream segment
% - fromIndices   : Index into link/segment variables (upstream indices)
% 
% DART Software - Copyright UCAR. Open source software, provided "as is" 
% Webpage: https://dart.ucar.edu/

filepath = '/Users/gharamti/Documents/MATLAB/Derecho/HydroDART/pywatershed/'; 
table_f  = 'river_midpoints.csv';
source_f = 'parameters_dis_seg.nc';
dest_f   = 'parameters_dis_seg_app.nc'; 

NTWRK_FILE = strcat(filepath, dest_f);     % Appended nc file
PARAM_FILE = strcat(filepath, source_f);   % Original params file
CSV_TABLE  = strcat(filepath, table_f);    % CSV table file

if ~isfile(PARAM_FILE) || ~isfile(CSV_TABLE)
    error([ 'Either the PARAM_FILE "' PARAM_FILE '" or the CSV_TABLE "' ...
          CSV_TABLE '" do not exist.' ])
end


%% Get lon from table (ArcGIS)
data = readtable(table_f);

inds = table2array(data(:, 5));
lats = table2array(data(:, 6));
lons = table2array(data(:, 7));

[~, isort] = sort(inds);

lats = lats(isort);
lons = lons(isort);

n_char = 15;

%% Compute Routelink style variables
ncPARAM = ncinfo(PARAM_FILE);

% read dimensions
n_seg = ncPARAM.Dimensions(1).Length;
gages = ncPARAM.Dimensions(2).Length;

% read variables
to_index   = double(ncread(PARAM_FILE, 'tosegment'));
poi_gauges = char(pad(ncread(PARAM_FILE, 'poi_gage_id'), n_char, 'left')); 

num_up_links  = zeros(1, n_seg);
upstream_seg  = struct;
fromIndsStart = zeros(1, n_seg);
fromIndsEnd   = zeros(1, n_seg);

k = 0;

for i = 1:n_seg
    
    % Upstream segments of i
    tmp = find(to_index == i)';

    % Find how many links are above me: 1, 2, 3, 4, ...
    % and get their indices. 
    num_up_links(i) = length(tmp);
    upstream_seg(i).links(:) = tmp;

    % If `fromIndsStart` and `fromIndsEnd` are the same 
    % and non-zero, that means there is only one link above me
    % num_up_links   = fromIndsEnd - fromIndsStart + 1; 
    % num_up_links(no_up_links) = 0;
    if ~isempty(tmp)
        fromIndsStart(i) = k+1;

        % Link i has upstream links
        for j = 1:num_up_links(i)
            k = k + 1;

            % Index into the state
            fromIndices(k) = tmp(j); %#ok
        end
        fromIndsEnd(i)   = k;
    end
end
n_fi = length(fromIndices);


%% Write NETCDF file
if isfile(NTWRK_FILE)
    warning([ 'A previosuly created "' dest_f '" file already exists. Deleting it ..' ])
    delete(NTWRK_FILE)
end

ncid = netcdf.create(NTWRK_FILE, 'NETCDF4'); % Full NetCDF-4 format

% Copy dimensions
for i = 1:length(ncPARAM.Dimensions)
    dim_name   = ncPARAM.Dimensions(i).Name;
    dim_length = ncPARAM.Dimensions(i).Length;
    
    % Use low-level netCDF function to define dimensions properly
    dim_ids.(dim_name) = netcdf.defDim(ncid, dim_name, dim_length);
end

% Define variables without data
var_ids = struct(); % Store variable IDs for reference
for i = 1:length(ncPARAM.Variables)
    var_name = ncPARAM.Variables(i).Name;
    var_dims = {ncPARAM.Variables(i).Dimensions.Name};  % Get dimension names
    var_type = ncPARAM.Variables(i).Datatype;

    % Convert dimension names to corresponding dimension IDs
    dim_id_list = [];
    for j = 1:length(var_dims)
        dim_id_list = [dim_id_list, dim_ids.(var_dims{j})]; %#ok
    end

    % Define the variable in the new NetCDF file
    var_ids.(var_name) = netcdf.defVar(ncid, var_name, var_type, dim_id_list);
end

% Exit definition mode
netcdf.endDef(ncid);

% Copy variable data
for i = 1:length(ncPARAM.Variables)
    var_name = ncPARAM.Variables(i).Name;
    data     = ncread(PARAM_FILE, var_name);
    
    % Write data to the new file
    netcdf.putVar(ncid, var_ids.(var_name), data);
end

nccreate(NTWRK_FILE, 'fromIndsStart', 'Dimensions', {'nsegment' , n_seg}, 'DataType', 'int64')
nccreate(NTWRK_FILE, 'fromIndsEnd'  , 'Dimensions', {'nsegment' , n_seg}, 'DataType', 'int64')
nccreate(NTWRK_FILE, 'fromIndices'  , 'Dimensions', {'index'    , n_fi }, 'DataType', 'int64')
nccreate(NTWRK_FILE, 'num_up_links' , 'Dimensions', {'nsegment' , n_seg}, 'DataType', 'int64')
nccreate(NTWRK_FILE, 'seg_lats'     , 'Dimensions', {'nsegment' , n_seg}, 'DataType', 'double')
nccreate(NTWRK_FILE, 'seg_lons'     , 'Dimensions', {'nsegment' , n_seg}, 'DataType', 'double')
nccreate(NTWRK_FILE, 'poi_gauges'   , 'Dimensions', {'IDLength' , n_char, 'npoigages', gages}, 'DataType', 'char')

ncwrite(NTWRK_FILE, 'fromIndsStart', fromIndsStart)
ncwrite(NTWRK_FILE, 'fromIndsEnd'  , fromIndsEnd)
ncwrite(NTWRK_FILE, 'fromIndices'  , fromIndices)
ncwrite(NTWRK_FILE, 'num_up_links' , num_up_links)
ncwrite(NTWRK_FILE, 'seg_lats'     , lats)
ncwrite(NTWRK_FILE, 'seg_lons'     , lons)
ncwrite(NTWRK_FILE, 'poi_gauges'   , poi_gauges')

netcdf.close(ncid); 
